concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -150,90 +251,186 @@ exportDatabaseV3 ::
+    --  - its genesis index = genindex and its height < startHeight
+    -- are previously exported and contained in chunks accounted
+    -- for in blockIndex. The file corresponding to the last chunk
+    -- entry of blockIndex contains the blocks at height > startHeight-1
+    -- at genesis index genIndex. This chunk contains no finalization
+    -- records. lastChunkFileM is the filename of the most recently
+    -- exported chunk containing all previously exported blocks of
+    -- genesis index == genIndex and height >= startHeight as well
+    -- as finalization records, if such a file exists.
+    (genIndex, startHeight, blockIndex, lastChunkFilenameM) <-
+        if indexFileExists
             then do
                 indexContents <- readFile indexFile
-                let lastChunkMetadata =
-                        if not $ null indexContents
-                            then splitOn "," . last . lines $ indexContents
-                            else -- if the index file is empty, reexporting from section 0 and block height 1 will be
-                            -- equivalent to exporting from scratch
-                                ["", "0", "1", "1"]
-                let lastChunkGenesisIndex = GenesisIndex . read $ lastChunkMetadata !! 1
-                let lastChunkFirstBlock = BlockHeight . read $ lastChunkMetadata !! 2
-                -- write back the index file without the last line (corresponding to the chunk deleted above)
-                writeFile (indexFile ++ "~") . unlines . init . lines $ indexContents
-                renameFile (indexFile ++ "~") indexFile
-                return (lastChunkGenesisIndex, lastChunkFirstBlock)
+                blockIndex <- case AP.parseOnly parseBlockIndexFile (T.pack indexContents) of
+                    Left err -> fail $ "An error occurred while parsing '" <> indexFile <> "': " <> err
+                    -- normalize the block index, s.t. superfluous genesis hash lines are removed.
+                    Right bi -> do
+                        return $ normalizeBlockIndex bi
+                case blockIndex of
+                    -- we export blocks starting from height 1 because genesis blocks need not be exported.
+                    Empty -> return (0, 1, Empty, Nothing)
+                    -- the following is safe due to normalization of the blockindex.
+                    blockIndexInit :|> (gd, chunks) -> case chunks of
+                        Empty -> fail $ "A section for genesis hash '" <> show gd <> "' in block index file has no entries."
+                        initChunks :|> BlockIndexChunkInfo{..} -> do
+                            -- the filename of the most recently exported chunk
+                            let lastChunkFile = outDir </> T.unpack filename
+                            -- if we overwrite files, back up the most recently exported
+                            -- chunk file so we can restore it if something goes wrong.
```

Comment:
I am not sure that this comment is particularly useful. If something goes wrong this program will likely crash, in which case the file will not be restored.